### PR TITLE
De-dup events in the base events queries (#74)

### PR DESCRIPTION
**Pull request by willbryant**
## Description & motivation

Fix duplicate event errors as discussed with @adamribaudo-velir and @dgitis in #74.

Previously we did this in the stg_ga4__events view, but that turned out to be incompatible with parition pruning, an important cost optimization.

Doing it in the "base" queries does not suffer from this problem because the incremental update process ensures that we are only querying one date partition at a time anyway, which is the right granularity to run de-duping at because the duplicate events have duplicate event_date_dt/event_timestamp values and therefore get delivered into the same partition.

## Checklist
- [x] I have verified that these changes work locally
- [x] I have updated the README.md (if applicable)
- [ ] I have added tests & descriptions to my models (and macros if applicable)
- [x] I have run `dbt test` and `python -m pytest .` to validate exists tests

Benchmarks below.

### CODE REVIEW

**Comment by willbryant** [comment]
TL;DR it is about 35-50% slower to build the base events, but it respects table pruning correctly so it processes no more data than the current main branch, and it fixes the duplicate events.

*Main branch*

Full run:

```
21:22:37  1 of 35 OK created incremental model test_dbt_ga4.base_ga4__events ............. [CREATE TABLE (116.5m rows, 106.8 GB processed) in 43.66s]
...
21:22:45  Completed with 1 error and 0 warnings:
21:22:45  
21:22:45  Failure in test unique_stg_ga4__events_event_key (models/staging/ga4/stg_ga4.yml)
21:22:45    Got 60038 results, configured to fail if != 0
```

Incremental run:

```
21:25:12  1 of 35 OK created incremental model test_dbt_ga4.base_ga4__events ............. [SCRIPT (3.3 GB processed) in 39.13s]
...
21:25:20  Completed with 1 error and 0 warnings:
21:25:20  
21:25:20  Failure in test unique_stg_ga4__events_event_key (models/staging/ga4/stg_ga4.yml)
21:25:20    Got 60038 results, configured to fail if != 0
```

Result:

```
SELECT count(*) FROM `<project>.test_dbt_ga4.base_ga4__events` 
116549295
```


*This branch*

Full run:

```
21:14:20  1 of 35 OK created incremental model test_dbt_ga4.base_ga4__events ............. [CREATE TABLE (116.5m rows, 106.8 GB processed) in 67.88s]
...
21:15:15  Completed successfully
```

Incremental run:

```
21:16:22  1 of 35 OK created incremental model test_dbt_ga4.base_ga4__events ............. [SCRIPT (3.3 GB processed) in 53.22s]
...
21:17:14  Completed successfully
```

Result:

```
SELECT count(*) FROM `<project>.test_dbt_ga4.base_ga4__events` 
116461189
```

**Comment by adamribaudo-velir** [comment]
This is awesome. I thought an approach like this would suffer from the same issues we see when using window functions in `stg_ga4__events`, but I think because we're incrementally loading in from a sharded table, as opposed to a partitioned table, we don't suffer the same issue with scanning costs. 

I just have 1 question before we merge: do you know how this behaves when `pseudo_user_id` and `ga_session_id` are null? Both are edge cases (pseudo_user_id is null when consent is denied and ga_session_id is null for audience triggers) so I'm fine merging, but wanted to give it some thought.  

My guess is that when either are null, records will be grouped into a single partition and we'll just pick the first.

**Comment by willbryant** [comment]
Yes, that's my understanding. So if two "anonymous" user events had the same exact timestamp with the same parameters, they would get combined to be one event (which may or may not be correct).

We could further reduce the likelihood of this occurring by also grouping by other fields. The origin of my particular list here is that that is the set of fields that goes into session_key & event_key, so it aligns with the uniqueness test we're going to run.

I suspect it's enough, but keen to hear others' thoughts.

**Comment by willbryant** [comment]
Thanks for doc updates!

**Comment by willbryant** [comment]
Incidentally can you ever get an audience trigger when the user_pseudo_id is null? I would have guessed that GA4 wouldn't be able to build an audience without giving the user a persistent identifier?

**Comment by adamribaudo-velir** [comment]
> Incidentally can you ever get an audience trigger when the user_pseudo_id is null?

Good call. Pretty certain I read that when consent is denied, they can't be a part of a GA4 audience. 

One last point I have about this PR is that deduping as part of a `base` model chafes against certain dbt conventions. I think we're backed into this approach due to how BQ operates, but before the next release we may want to rename the model or otherwise add documentation to clarify the rationale.

**Review by adamribaudo-velir — APPROVED** [review]

**Comment by willbryant** [comment]
Yeah, I would ideally like to address #80 and what I have in mind there would also address #33 - I'd like to made the "base" layer a single query on `events_*` and do the persisting-or-not in the next layer after that, with the same _table_suffix filtering mechanism in both the persisted daily and the non-persisted intraday/since-last-persisted queries.

Probably clearer in a PR than my explanation :)
